### PR TITLE
Add resiliency around missing scheduling info.

### PR DESCRIPTION
**Pull request by jeffchao**
### Context

This commit uses work from Netflix/mantis#37 to
address edge cases for missing scheduling info.

We catch the following cases:

1. On Mantis Master failover.
  - There are probably existing running workers but since it's a
newly-elected leader, it needs to update its view of the world. If Mesos
doesn't send it legit worker information, **mark them for resubmission and
resubmit them in bulk at the end of initialization**.
  - If it fails to load worker info from persistence due to corruption,
**proactively resubmit the worker**.
2. On a fresh job submit.
  - Master will notice that the worker doesn't have the correct
information. **Proactively unschedule the worker for the assignment
result**.

This commit also adds one metric:

1. `numMissingWorkerPorts` which measures the number of workers
with missing ports, specifically on master failover.

Lastly, this commit does some minor refactoring:

1. Some helper methods.
2. Javadocs clarity.
3. **Fix broken tests due to tighter validation around `WorkerPorts`**.
  - Also makes sure our tests use the same `WorkerPorts` constructor signature
as the real service.

### Checklist

- [x] `./gradlew build` compiles code correctly
- [x] Added new tests where applicable
- [x] `./gradlew test` passes all tests
- [x] Extended README or added javadocs where applicable
- [x] Added copyright headers for new files from `CONTRIBUTING.md`

### CODE REVIEW

File: server/src/main/java/io/mantisrx/master/jobcluster/WorkerInfoListHolder.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/master/jobcluster/WorkerInfoListHolder.java:44]: We weren't showing this in log output.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: Need a closer review here. This is on failover. Idea is to mark bad workers one by one. Then, at the end of this init, resubmit them all together.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: Need a review here too. The previous boolean condition is now delegating logic to the `MantisWOrkerMetadataImpl`. It does the same thing, except it returns a list of ports (good or bad) instead of an empty list. Intent is to not get array index out of bounds exception.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java]: Need a closer review here. Idea is to make sure worker metadata is good even on a heartbeat. If not, bubble up this exception.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java]: Building this list exactly as it previously was, just moved this method to this spot.

File: server/src/main/java/io/mantisrx/server/master/SchedulingService.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/server/master/SchedulingService.java:364]: Need a closer review here. Intent is to detect bad worker ports when launching the task. If bad, then don't even send a `WorkerLaunched` event. Expect this worker to be stuck in a state to be resubmitted in the normal stuck/resubmit workflow.

File: server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java
Comment (jeffchao) [inline server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java:422]: For this test, and other tests, they were initially failing. This was because the tests use a different constructor than `SchedulingService`. We can update the `WorkerPort` constructor(s), but they aren't being used. It's better to make the usage match what the scheduler is really using.

**Comment by jeffchao** [comment]
@codyrioux @neerajrj @nickmahilani @piygoyal submitting this PR for some early feedback. I commented in a few places for things I'd like a closer review/second opinion. Initial testing looks good. I'll continue with more rigorous testing tomorrow. Also check out the commit/PR description for more context. Thanks.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: Also narrowed the exception down to catch just the 3 we need.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java
Comment (nickmahilani) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java]: this part of the comment is unnecessary IMHO, the git history should have this context.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java
Comment (nickmahilani) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java]: feels odd to have this check here. The validation should ideally happen on state changes at a central place. A worker goes from `Accepted` to `Launched` (only when there are valid ports after scheduling) to `StartInitiated` (worker lifecycle start) to `Started` (when there is a heartbeat). Are we seeing a HB before the worker is in `Launched` state?
If we can ensure the WorkerMetadata is in a consistent state with valid port info if it is in any of Launched/StartInitiated/Started states at the time of transitioning to those states, we might be able to avoid these validity checks at multiple places.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java
Comment (nickmahilani) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java]: since WorkerMetadata already has a getWorkerPorts(), this method feels redundant. The calling code can derive this info from getWorkerPorts()?

File: server/src/main/java/io/mantisrx/server/master/SchedulingService.java
Comment (nickmahilani) [reply server/src/main/java/io/mantisrx/server/master/SchedulingService.java:364]: think we could proactively unschedule the worker here if the assignmentResult doesn't return the expected number of ports.. Relying on standard flow for resubmit would wait for heartbeat timeouts which could take a while

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java]: Agreed. I wasn't sure about the HB sending. Let me remove this and test out the HB and state transitions.

Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java]: To clarify, originally Neeraj and I were thinking to catch it here but it's worth testing out situation you pointed out.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java]: Sounds good, I'll remove.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/MantisWorkerMetadataImpl.java]: I think I can move this out of this class. The important thing is that we don't return an empty list as we did before https://github.com/Netflix/mantis-control-plane/pull/26/files#diff-81433f6ab0cca184eca4716606d9b3eeL1448-L1455. From my understanding, we don't have ports in the data structure initially. Port info is filled in afterwards, so methods like `wm.getMetricsPort()` return a `-1` if worker ports don't exist. I can move this out and build a list of `-1`s in `JobActor`. thoughts?

File: server/src/main/java/io/mantisrx/server/master/SchedulingService.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/server/master/SchedulingService.java:364]: Makes sense. I'll proactively unschedule if there's no worker ports.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/worker/JobWorker.java]: Thinking this through some more, I think we can do this without going through heartbeats. in `JobActor#initializeRunningWorkers` https://github.com/Netflix/mantis-control-plane/pull/26/files#diff-81433f6ab0cca184eca4716606d9b3eeL1363, we can just check for missing ports without interfering with the fake heartbeat send. Then, we can get rid of the exception handling. cc @neerajrj

**Comment by jeffchao** [comment]
@neerajrj can you do a pass at review? tested failover case with single worker job, multiple worker job (multi worker fail, all worker fail) and it looks good. The PR description has the most relevant context/explanation.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (neerajrj) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: A nitpick but adding job id to the error log may be helpful as the generic stack trace may not reveal where the issue occurred

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (neerajrj) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: lgtm

File: server/src/main/java/io/mantisrx/server/master/SchedulingService.java
Comment (neerajrj) [inline server/src/main/java/io/mantisrx/server/master/SchedulingService.java:368]: Since this block is hit primarily for invalid /missing ports, should we name the counter appropriately so we can track all invalid port related issues together?

File: server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java
Comment (neerajrj) [reply server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java:422]: Are there any unit tests for the invalid ports on master restart case?

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: Agreed. This was left from before but I'll go ahead and change it.

File: server/src/main/java/io/mantisrx/server/master/SchedulingService.java
Comment (jeffchao) [reply server/src/main/java/io/mantisrx/server/master/SchedulingService.java:368]: Yeah that works for me. I'll use that same counter `numMissingWorkerPorts`.

File: server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java
Comment (jeffchao) [inline server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: @nickmahilani @neerajrj a last iteration:

I'm explicitly marking the worker as `FAILED` if there's no ports. When the worker gets loaded, the scheduler will notice that it's in a bad state and resubmit. With this approach, we don't have to do anymore weird Null checking before sending scheduling info.

File: server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java
Comment (nickmahilani) [inline server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: this would also resubmit Accepted/Failed state workers? The worker ports validation is only applicable to running workers.

Comment (jeffchao) [reply server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java]: I see. I'll move this into the block below `WorkerState.isRunningState`.

File: server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java
Comment (nickmahilani) [reply server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: Worker ports should be checked only for workers in Launched/StartInitiated/Started state. Can this validation be moved to inside the JobWorker.build method where it does other set of validations?

Comment (jeffchao) [reply server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: It would, but also `Accepted`. From your other suggestion, I'll change this to only use Launched/StartInitiated/Started via `MantisJobState.isRunningState`. Unit test for this to come shortly.

File: server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java
Comment (nickmahilani) [inline server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: can also add a JobcompletedReason when marking it FAILED.. 
Maybe, even add a metric that we could alert on it if this happens.

Comment (jeffchao) [reply server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: I thought about that and setting `completedAt` as well. But is the job really completed? There could be other workers which could be fine.

Comment (nickmahilani) [reply server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: The enum is confusing due to legacy reasons, it is used for both worker and job IIRC. In this case, it only applies to the worker and not the job.. If it is making this PR complex, feel free to leave it out and address later

Comment (jeffchao) [reply server/src/main/java/io/mantisrx/server/master/domain/DataFormatAdapter.java]: Oh interesting. No problem, adding completed reason (`Killed`) is a simple change. I'll do that.

**Comment by jeffchao** [comment]
Got a last review from Neeraj. Looks good. Merging.

**Review by neerajrj — APPROVED** [review]
Looks good!

**Comment by jeffchao** [comment]
@neerajrj yeah we've got invalid port tests from https://github.com/Netflix/mantis/blob/master/mantis-common/src/test/java/io/mantisrx/common/WorkerPortsTest.java.

File: server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java
Comment (jeffchao) [reply server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java:422]: @neerajrj yeah we've got invalid port tests from https://github.com/Netflix/mantis/blob/master/mantis-common/src/test/java/io/mantisrx/common/WorkerPortsTest.java. Also for the master restart: https://github.com/Netflix/mantis-control-plane/blob/master/server/src/test/java/io/mantisrx/master/jobcluster/job/JobClusterManagerTest.java#L624.

**Pull request merged**
